src/builders/thermodynamic.jl

```diff
@@ -0,0 +1,258 @@
+        for met in keys(A.reaction_stoichiometry(model, rid))
+            push!(mets, Symbol(met))
+        end
+    end
+    mets = [m for m in mets] # TODO: constrainttrees #12
+    stoi = A.stoichiometry(model)
+
+    # create thermodynamic variables
+    m = C.ConstraintTree(:max_min_driving_force^C.variable())
+    m +=
+        :log_metabolite_concentrations^C.variables(
+            keys = mets,
+            bounds = zip(log(concentration_lb), log(concentration_ub)),
+        )
+    m += :delta_G_reactions^C.variables(keys = Symbol.(rxns))
+
+    #= 
+    Build gibbs free energy relations
+    ΔG_rxn == ΔG0 + R * T * sum ( log_concentration_variable * stoichiometry_value )
+    =#
+    model_rxns = A.reactions(model)
+    model_mets = A.metabolites(model)
+    for rxn in rxns
+        met_idxs, stoich_coeffs = findnz(stoi[:, findfirst(==(rxn), model_rxns)])
+        met_ids = model_mets[met_idxs]
+
+        dG0 = reaction_standard_gibbs_free_energies[rxn]
+
+        m *=
+            :delta_G_reaction_equations^Symbol(
+                rxn,
+            )^C.Constraint(
+                value = -m.delta_G_reactions[Symbol(rxn)].value +
+                        dG0 +
+                        R *
+                        T *
+                        sum(
+                            m.log_metabolite_concentrations[Symbol(met_id)].value *
+                            stoich for (met_id, stoich) in zip(met_ids, stoich_coeffs)
+                        ),
+                bound = 0.0,
+            )
+    end
+
+    #= 
+    Set proton log concentration to zero so that it won't impact any
+    calculations (biothermodynamics assumption). Also set water concentrations
+    to zero (aqueous condition assumptions). How true is "aqueous conditions"? 
+    Debatable... 
+    =#
+    log_met_conc_zero = [Symbol.(proton_ids); Symbol.(water_ids)]
+    for met in keys(m.log_metabolite_concentrations)
+        if met in log_met_conc_zero
+            m.log_metabolite_concentrations[met].bound = 0.0
+        end
+    end
+
+    #= 
+    Add thermodynamic feasibility constraint (ΔG < 0 for a feasible reaction in flux direction).
+    Add objective constraint to solve max min problem.
+    =#
+    for rxn in rxns
+        m *=
+            :thermodynamic_feasibility^Symbol(
+                rxn,
+            )^C.Constraint(
+                value = m.delta_G_reactions[Symbol(rxn)].value *
+                        sign(get(flux_solution, rxn, 1.0)),
+                bound = (-Inf, 0.0),
+            )
+
+        m *=
+            :max_min_constraints^Symbol(
+                rxn,
+            )^C.Constraint(
+                value = m.max_min_driving_force.value +
+                        m.delta_G_reactions[Symbol(rxn)].value *
+                        sign(get(flux_solution, rxn, 1.0)),
+                bound = (-Inf, 0.0),
+            )
+    end
+
+    m
+end
+
+export build_max_min_driving_force_model
+
+function add_metabolite_ratio_constraints!(
```

Comment:
this does not allocate variables, so there's absolutely no reason for the inplace functionality.

Also you can generalize it to ratio_constraints; it may constraint anything (even fluxes)

Reply:
This is so much better!